application/src/main/java/org/thingsboard/server/service/entitiy/ruleChain/DefaultTbRuleChainNotifyService.java

```diff
@@ -0,0 +1,266 @@
+
+            for (RuleChain updatedRuleChain : updatedRuleChains) {
+                RuleChainMetaData updatedRuleChainMetaData = checkNotNull(ruleChainService.loadRuleChainMetaData(tenantId,
+                        updatedRuleChain.getId()));
+                notificationEntityService.notifySaveRuleChainMetaData(tenantId, ruleChain, updatedRuleChain,
+                        updatedRuleChainMetaData, user);
+            }
+
+            return savedRuleChainMetaData;
+        } catch (Exception e) {
+            notificationEntityService.notifyCreateOrUpdateOrDelete(tenantId, null, emptyId(EntityType.RULE_CHAIN),
+                    null, user, ActionType.ADDED, false, e,  ruleChainMetaData);
+            throw handleException(e);
+        }
+    }
+
+    @Override
+    public RuleChain assignRuleChainToEdge(TenantId tenantId, RuleChainId ruleChainId, Edge edge, SecurityUser user) throws ThingsboardException {
+        try {
+            RuleChain savedRuleChain = checkNotNull(ruleChainService.assignRuleChainToEdge(tenantId, ruleChainId, edge.getId()));
+            notificationEntityService.notifyAssignOrUnassignEntityToEdge(tenantId, ruleChainId,
+                    null, edge.getId(),
+                    savedRuleChain, ActionType.ASSIGNED_TO_EDGE,
+                    user, ruleChainId.toString(), edge.getId().toString(), edge.getName());
+            return savedRuleChain;
+        } catch (Exception e) {
+            notificationEntityService.notifyCreateOrUpdateOrDelete(tenantId, null, emptyId(EntityType.RULE_CHAIN),
+                    null, user, ActionType.ASSIGNED_TO_EDGE, false, e, ruleChainId.toString(), edge.getId().toString());
+            throw handleException(e);
+        }
+    }
+
+    @Override
+    public RuleChain unassignRuleChainFromEdge(TenantId tenantId, RuleChain ruleChain, Edge edge, SecurityUser user) throws ThingsboardException {
+        RuleChainId ruleChainId = ruleChain.getId();
+        EdgeId edgeId = edge.getId();
+        try {
+            RuleChain savedRuleChain = checkNotNull(ruleChainService.unassignRuleChainFromEdge(tenantId, ruleChainId, edgeId, false));
+            notificationEntityService.notifyAssignOrUnassignEntityToEdge(tenantId, ruleChainId,
+                    null, edge.getId(),
+                    savedRuleChain, ActionType.UNASSIGNED_FROM_EDGE,
+                    user, ruleChainId.toString(), edge.getId().toString(), edge.getName());
+            return savedRuleChain;
+        } catch (Exception e) {
+            notificationEntityService.notifyCreateOrUpdateOrDelete(tenantId, null, emptyId(EntityType.RULE_CHAIN),
+                    null, user, ActionType.UNASSIGNED_FROM_EDGE, false, e, ruleChainId.toString(), edge.getId().toString());
+            throw handleException(e);
+        }
+    }
+
+    @Override
+    public RuleChain setEdgeRuleChain(TenantId tenantId, RuleChain ruleChain, SecurityUser user, boolean isSetEdgeTemplateRootRuleChain) throws ThingsboardException {
```

Comment:
please split it into two different methods:
**setEdgeTemplateRootRuleChain** and **setAutoAssignToEdgeRuleChain**

The current implementation with additional boolean variable adds complexity to the code reading.

Reply:
Ok